# Cafe Sales Analytics Projects

## Objectives
Audit the raw datasets of Cafe Sales to understand the structure, quality issues and potential cleaning steps to take.

In [2]:
import pandas as pd
import numpy as np

In [3]:
import os
os.getcwd()

'/home/jv546/data_projects/cafe_analytics_project/notebooks'

In [4]:
#import CSV
data_path = "../data/raw/dirty_cafe_sales.csv"
try:
    df = pd.read_csv(data_path)
except FileNotFoundError:
    print("ERROR : File Not Found")
pd.reset_option("display.max_row")

In [5]:
df

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
2       TXN_4271903    Cookie        4            1.0       ERROR   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9995    TXN_7672686    Coffee        2            2.0         4.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9997    TXN_5255387    Coffee        4            2.0         8.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07  

[10000 rows x 8 columns]

## Check Structure

In [6]:
df.shape

(10000, 8)

In [7]:
df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


## Sample Data

In [9]:
df.head(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2            2.0         4.0   
1    TXN_4977031      Cake        4            3.0        12.0   
2    TXN_4271903    Cookie        4            1.0       ERROR   
3    TXN_7034554     Salad        2            5.0        10.0   
4    TXN_3160411    Coffee        2            2.0         4.0   
5    TXN_2602893  Smoothie        5            4.0        20.0   
6    TXN_4433211   UNKNOWN        3            3.0         9.0   
7    TXN_6699534  Sandwich        4            4.0        16.0   
8    TXN_4717867       NaN        5            3.0        15.0   
9    TXN_2064365  Sandwich        5            4.0        20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

In [10]:
df.tail(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
9990    TXN_1538510    Coffee        5            2.0        10.0   
9991    TXN_3897619  Sandwich        3            4.0        12.0   
9992    TXN_2739140  Smoothie        4            4.0        16.0   
9993    TXN_4766549  Smoothie        2            4.0         NaN   
9994    TXN_7851634   UNKNOWN        4            4.0        16.0   
9995    TXN_7672686    Coffee        2            2.0         4.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9997    TXN_5255387    Coffee        4            2.0         8.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
9990  Digital Wallet       NaN       2023-05-22  
9991            Cash  Takeaway       2023-02-24  
9992         UNKNOWN  In-store       2023-07-05  
9993            Cash       NaN       2023-10-20  
9994             NaN       NaN       2023-01-08  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07

## Check Data Quality

In [15]:
df.isnull().sum()

Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

In [ ]:
(df.isnull().sum()/ len(df))*100

Transaction ID       0.00
Item                 3.33
Quantity             1.38
Price Per Unit       1.79
Total Spent          1.73
Payment Method      25.79
Location            32.65
Transaction Date     1.59
dtype: float64

In [ ]:
df.duplicated().sum()

np.int64(0)

## Distribution Analysis

In [14]:
df.describe(include="all")

Transaction ID   Item Quantity Price Per Unit Total Spent  \
count           10000   9667     9862           9821        9827   
unique          10000     10        7              8          19   
top       TXN_9226047  Juice        5            3.0         6.0   
freq                1   1171     2013           2429         979   

        Payment Method  Location Transaction Date  
count             7421      6735             9841  
unique               5         4              367  
top     Digital Wallet  Takeaway          UNKNOWN  
freq              2291      3022              159

## Initial Observations
- This Dataset have 10 000 rows with 8 columns 
- no duplicate data
- unique transaction id with no missing or invalid data
- all column have the type object (needs to be fixed)

## Potential Issues
- there is quite a few null values per column as a percentage 
    Item                 3.33%
    Quantity             1.38%
    Price Per Unit       1.79%
    Total Spent          1.73%
    Payment Method      25.79%
    Location            32.65%
    Transaction Date     1.59%
- There is also a bunch of unknown and error values that need to be fixed for each colmn.
- missing values in the price per unit column needs to be verifyed and correct to the propre amount
- verify total spent calculations using Quantity * Price Per Unit.
- determining missing quantity by reverse calculating using the total and price per unit
- set transaction dates to datetime and order them
- fix data types for each columns

## Questions to Resolves
-TODO